### PR TITLE
[FEATURE] Added Fast Url Rewrites functionality so no category produc…

### PR DESCRIPTION
**Pull request by lewisvoncken**
…t urls are generated when this is disabled in the configuration

### Description

When the SEO setting _Use Categories Path for Product URLs_ is disabled this functionality will make sure that these urls are not generated so the process of creating url rewrites will be more efficient.

### Manual testing scenarios
The following testing scenario is in a default installation in which the _Use Categories Path for Product URLs_ is set to **No**

1. Create a Product
2. Create a Category
3. Assign the Product to the Category
4. Check the url_rewrites table which will now show 4 rewrites which refer to the category and the product
5. Remove these rewrites
6. Apply Changes
7. Update the url key for the category
8. Update the url key for the product
10. Now there are only shown 2 rewrites which are necessary.

Example:
image

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: app/code/Magento/CatalogUrlRewrite/Observer/UrlRewriteHandler.php [hunk 1]
```diff
@@ -70,7 +76,8 @@ public function __construct(
         \Magento\Catalog\Model\ResourceModel\Product\CollectionFactory $productCollectionFactory,
         \Magento\CatalogUrlRewrite\Model\CategoryProductUrlPathGenerator $categoryBasedProductRewriteGenerator,
         \Magento\UrlRewrite\Model\MergeDataProviderFactory $mergeDataProviderFactory = null,
-        \Magento\Framework\Serialize\Serializer\Json $serializer = null
+        \Magento\Framework\Serialize\Serializer\Json $serializer = null,
+        \Magento\Framework\App\Config\ScopeConfigInterface $scopeConfig
```
Comment (ishakhsuvarov) [inline app/code/Magento/CatalogUrlRewrite/Observer/UrlRewriteHandler.php:80]: A new required argument can not be added after optional ones

**Review by ishakhsuvarov — CHANGES_REQUESTED** [review]

**Comment by ishakhsuvarov** [comment]
Hi @lewisvoncken 
Thank you for your Pull Request. Could you please check failing integration tests?

Thanks

**Comment by hostep** [comment]
Even though I completely understand the need for this, I think it's a very bad idea because it will break when people change back the setting "Use Categories Path for Product URLs" to "Yes".

Take this scenario (tested with Magento 2.2.4 and the changes proposed in this PR applied on top of that codebase):

1. Start from a clean installed Magento database
2. Create a category and make sure its url key is 'category'
3. Create a product and make sure its url key is 'product'
4. Reindex & flush caches
5. Go to the frontend and navigate to the Category & Product and look at the urls, result:
- https://example.com/category.html
- https://example.com/product.html

=> Good is expected

6. In the backend, change the configuration "Use Categories Path for Product URLs" to "Yes"
7. Reindex & flush caches
8. Go to the frontend and navigate to the Category & Product and look at the urls, result:
- https://example.com/category.html
- https://example.com/catalog/product/view/id/1/s/product/category/3/

=> Not good, the product url should be: https://example.com/category/product.html

As long as Magento doesn't have a mechanism to regenerate url rewrites, I don't think this PR should get merged.

Thanks!

**Comment by ishakhsuvarov** [comment]
Hi @lewisvoncken 
Comment above by @hostep seems to be completely valid. Do you have any idea how to improve this behavior?
Thanks

**Comment by ishakhsuvarov** [comment]
@lewisvoncken Closing this PR due to inactivity and concerns raised in the comments above.
Please feel free to reopen if you wish to update

**Pull request closed without merging**